migration/migration.go

```diff
@@ -73,30 +89,41 @@ func MigrateIfNeeded(targetDB db.DB, network utils.Network, log utils.SimpleLogg
 		new ones. It will be able to do this since the schema version it reads from the database will be
 		non-zero and that is what we use to initialise the i loop variable.
 	*/
-	version, err := SchemaVersion(targetDB)
+	metadata, err := SchemaMetadata(targetDB)
 	if err != nil {
 		return err
 	}
 
-	for i := version; i < uint64(len(migrations)); i++ {
+	for i := metadata.Version; i < uint64(len(migrations)); i++ {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
 		log.Infow("Applying database migration", "stage", fmt.Sprintf("%d/%d", i+1, len(migrations)))
 		migration := migrations[i]
-		migration.Before()
+		err := migration.Before(metadata.IntermediateState)
+		if err != nil {
+			return err
+		}
 		for {
 			var migrationErr error
 			if dbErr := targetDB.Update(func(txn db.Transaction) error {
-				migrationErr = migration.Migrate(txn, network)
-				if migrationErr != nil {
-					if errors.Is(migrationErr, ErrCallWithNewTransaction) {
-						return nil // Run the migration again with a new transaction.
-					}
+				metadata.IntermediateState, migrationErr = migration.Migrate(ctx, txn, network)
+				switch migrationErr {
+				case ctx.Err():
+				case nil:
+				case ErrCallWithNewTransaction:
+					return nil // Run migration again with new transaction.
+				default:
```

Comment:
Now that I think about it, we can simplify this even further:
```go
if migrationErr != nil {
        if errors.Is(migrationErr, ErrCallWithNewTransaction) {
                return nil // Run migration again with new transaction.
        }
        return migrationErr
}
```
This is closer to what you had originally (with one less branch I think). Apologies for the back-and-forth.

Reply:
The metadata won't be updated if ctx is canceled

Reply:
Ah you're right, never mind.